Small updates and tweaks after a from-scratch setup

cpu opened the pull request:
This branch adds a handful of changes (some discussed out-of-band) to the project based on my experience setting this up on a test server from scratch.

The `flake.nix`, `flake.lock` and `.envrc` changes might be controversial. If you'd prefer I keep that config out-of-repo I can manage, but since it sounds like the expectation is that I'll be the primary maintainer of this code going forward I'd prefer to make life easier and include it :-)

cpu commented on ansible/roles/build/tasks/main.yml:
```diff
@@ -0,0 +1,28 @@
+- name: 'Git checkout'
+  become_user: '{{app_user}}'
+  git:
+    repo: 'https://github.com/cpu/rustls-bench-app/'
+    dest: /home/{{app_user}}/rustls-bench-app
+    depth: 1
+    version: cpu-review
```
oops: need some vars here to not use my repo/branch :) Will fix

cpu commented:
Success :tada: - https://github.com/cpu/rustls/pull/4#issuecomment-1789470200

I had two small configuration issues, both addressed in this branch:

1. The GitHub app needs to configure the webhook URL with the correct route path: 7ea9d5ac51c8031e857434a80af0a4b8c3a087e1
2. The benchmarking server needs `valgrind` installed for `cachegrind`: c512d0cd2ef44c79b6784c3ef89e11173875c22e

I also noticed a couple small things we might want to dig in to further:

1. Before I had `cachegrind` available the run failed, and so the GitHub app left a nice helpful comment with the log results. I deleted that comment after fixing the root cause, but when I bumped the Bot in a comment it entered into a crash loop with errors like:
```
Nov 01 18:35:11 rustls-bench bash[42289]: error handling issue comment: GitHub: Not Found
Nov 01 18:35:11 rustls-bench bash[42289]: Documentation URL: https://docs.github.com/rest/issues/comments#update-an-issue-comment
Nov 01 18:35:11 rustls-bench bash[42289]: Caused by:
Nov 01 18:35:11 rustls-bench bash[42289]:     Not Found
Nov 01 18:35:11 rustls-bench bash[42289]:     Documentation URL: https://docs.github.com/rest/issues/comments#update-an-issue-comment
```
I had to stop the service and restart it for it to recover. I think it was trying to update the comment that I deleted.

2. We're ignoring PR events for "reopened". I think it might make sense to run benchmarking for that event too. WDYT?

cpu commented on ansible/roles/deploy/tasks/main.yml:
```diff
@@ -47,3 +38,6 @@
     group: 'root'
   notify:
     - 'start app'
+- name: 'Install valgrind'
+  apt:
+    name: valgrind
```
I wasn't sure if this is the best place to add this.

Maybe the whole `deploy` task should be renamed now that the service binary is built on-host and this task is mostly setting up the systemd unit and config?

aochagavia replied:
I think installing valgrind here makes sense, by lack of a better place (or we could rename the `cpufreq` role to something else like `misc` and put dependencies like valgrind and git there).

If we keep valgrind here, I'd suggest making it the first step of the task, to ensure valgrind is installed before the app runs (it probably doesn't make a difference 99% of the time, but feels more correct and might save someone down the line a confusing error)

Regarding the name of the `deploy` roole, I'm having trouble thinking of something more suitable... Maybe `bench-app` would do. Or did you have another name in mind?

cpu replied:
> If we keep valgrind here, I'd suggest making it the first step of the task

Makes sense to me :+1: 

> Maybe bench-app would do. Or did you have another name in mind?

That seems like a good name to me :-)

cpu replied:
> If we keep valgrind here, I'd suggest making it the first step of the task

Fixed.

> Maybe bench-app would do. Or did you have another name in mind?

I went with `bench-app-service`.

aochagavia commented on ansible/roles/deploy/templates/config.json:
```diff
@@ -3,10 +3,10 @@
-  "github_repo_owner": "rustls",
-  "github_repo_name": "rustls",
+  "github_app_id": {{ github_app_id }},
+  "github_app_key": "{{ github_app_key.replace('\n', '\\n') }}",
```
Smart, I didn't think of this and manually removed the newlines from my private key (somehow the PEM parser didn't complain)

aochagavia commented on ansible/secrets.example.yml:
```diff
@@ -1,8 +1,17 @@
-github_app_key: '-----BEGIN RSA PRIVATE KEY-----<base64_encoded_key>-----END RSA PRIVATE KEY-----'
+github_app_key: |-
+  -----BEGIN RSA PRIVATE KEY-----
+  <SECRET BASE46 GOES HERE>
```
Nitpick: it should say BASE64 instead of BASE46

cpu replied:
Oops :laughing: Will fix.

cpu replied:
Fixed!

aochagavia commented on ci-bench-runner/Cargo.toml:
```diff
@@ -12,7 +12,7 @@ hyper = { version = "0.14.27", default-features = false }
 # TODO: replace by a crates.io version once they publish it
 octocrab = { git = "https://github.com/XAMPPRocky/octocrab.git" }
-sentry = { version = "0.31.7", features = ["tracing"] }
+sentry = { version = "0.31.7", features = ["tracing", "ureq", "rustls"], default-features = false }
```
Out of curiosity, what does this do?

cpu replied:
I added more of this information to the commit message.

cpu commented on ansible/readme.md:
````diff
@@ -41,10 +45,24 @@ too old.
+  --extra-vars 'bench_app_repo=https://github.com/example/rustls-bench-app/' \
+  --extra-vars 'bench_app_branch=example-branch' \
+  playbook.yml
+```
````
> Is this about testing that the Ansible bits are working properly, or also about testing that the application works after the deploy? In the second case, I think a user would need a bit more guidance by mentioning things like:

I was thinking a bit of both! Mainly I was looking ahead to the days where we have this code running in a "production" deploy on the dedicated server. If we need to make any substantive changes to the application code or to the ansible scripts we'll probably want to test it without impacting the production deploy. I don't think it makes sense to have a whole second server always laying around for this testing, but it would be helpful if we had instructions for setting one up on-demand.

> In the second case, I think a user would need a bit more guidance by mentioning things like:

These are good points. I'll try to add some more text with the points you describe. I think having them written down in the README or somewhere equivalent will be helpful long term.

> It almost feels like we should have a terraform script to create the GitHub forks, the application and its installation 😛... But let's not go down that path.

Haha yeah I think that's probably overkill for now :fearful: (but would be cool!)

cpu replied:
> These are good points. I'll try to add some more text with the points you describe.

I updated the testing section of the ansible README to cover the motivation, and the required up front setup.

cpu commented on ansible/roles/build/tasks/main.yml:
```diff
@@ -0,0 +1,28 @@
+      group: '{{app_user}}'
+      mode: 'u=rwx,g=rx,o=rx'
+  notify:
+    - 'start app'
```
Agreed :)

cpu commented on ansible/playbook.yml:
```diff
@@ -35,6 +38,11 @@
+    import_role:
+      name: 'build'
+    tags:
+      - 'build'
```
I agree with you that running just the build is probably not very useful in practice. I'll combine the tags :+1:

cpu commented:
> I like them! Could you maybe explain in the root readme how someone might benefit from this setup? 

Sure, can do :+1: 

> We need to gracefully fall back to creating a new comment when the original comment cannot be updated / no longer exists (edit: implemented in https://github.com/aochagavia/rustls-bench-app/commit/8753925980a7661197697c22e5c521f24d61b9b6).

Nice work!

> Do you have an opinion on this? I'm tempted to go back to the "no retry" camp, to keep things simple (also, I discovered that octocrab automatically retries GitHub requests when HTTP status indicates a transient failure).

I think that makes sense, especially in the early days. We can kick the bot into retrying manually with a comment or closing/opening the PR, right?

> I was expecting that might be covered by the "synchronize" event action, but we can definitely add "reopened" too (edit: added in https://github.com/aochagavia/rustls-bench-app/commit/bd9ce252faebbab9820d4f43371bd8c156f27952).

Great, thanks again.

I will push revisions based on your feedback shortly.

cpu commented on ansible/readme.md:
```diff
@@ -4,8 +4,11 @@ This directory contains the necessary scripts for the bench server's setup and m
+
+* securing the server after setting up the operating system and creating users for server admins is not included.
+* performing a deploy _will_ affect benchmarking results (for example due to compiling Rust code for `ci-bench-runner`).
+  Try to avoid deploying while benchmarking is in progress, or re-run benchmarks afterwards to avoid noise.
```
Fixed!

cpu commented:
> Could you maybe explain in the root readme how someone might benefit from this setup?

I added some brief nix specific info to the top-level README.

> I will push revisions based on your feedback shortly.

I rebased on main and addressed all of your feedback. Thanks for the review.

aochagavia commented:
This is great, thanks! And now I'll roll up my sleeves to try out nix, at last.
